test/system/src/test/java/io/pravega/test/system/ReadWriteAndAutoScaleWithFailoverTest.java

```diff
@@ -0,0 +1,209 @@
+
+        // Fetch all the RPC endpoints and construct the client URIs.
+        final List<String> uris = conURIs.stream().filter(uri -> uri.getPort() == 9092).map(URI::getAuthority)
+                .collect(Collectors.toList());
+
+        controllerURIDirect = URI.create("tcp://" + String.join(",", uris));
+        log.info("Controller Service direct URI: {}", controllerURIDirect);
+
+        // Verify segment store is running.
+        segmentStoreInstance = new PravegaSegmentStoreService("segmentstore", zkUri, controllerURIDirect);
+        assertTrue(segmentStoreInstance.isRunning());
+        log.info("Pravega Segmentstore service instance details: {}", segmentStoreInstance.getServiceDetails());
+
+        //executor service
+        executorService = Executors.newFixedThreadPool(NUM_READERS + TOTAL_NUM_WRITERS);
+        //get Controller Uri
+        controller = new ControllerImpl(controllerURIDirect);
+        testState = new TestState();
+    }
+
+    @After
+    public void tearDown() {
+        controllerInstance.scaleService(1, true);
+        segmentStoreInstance.scaleService(1, true);
+        executorService.shutdownNow();
+        testState.eventsReadFromPravega.clear();
+    }
+
+
+    @Test(timeout = 600000)
+    public void readWriteAndAutoScaleWithFailoverTest() throws Exception {
+        createScopeAndStream(scope, STREAM, config, controllerURIDirect);
+        log.info("Scope passed to client factory {}", scope);
+        try (ClientFactory clientFactory = new ClientFactoryImpl(scope, controller);
+             ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(scope, controllerURIDirect)) {
+
+            createReadersAndWriters(clientFactory, readerGroupName, scope, readerGroupManager, STREAM, INIT_NUM_WRITERS, NUM_READERS);
+
+            //run the failover test before scaling
+            performFailoverTest();
+
+            //bring the instances back to 3 before performing failover during scaling
+            controllerInstance.scaleService(3, true);
+            segmentStoreInstance.scaleService(3, true);
+            Thread.sleep(ZK_DEFAULT_SESSION_TIMEOUT);
+
+            addNewWriters(scope, clientFactory);
+
+            //run the failover test while scaling
+            performFailoverTest();
+
+            //wait for scaling
+            waitForScaling();
+
+            //bring the instances back to 3 before performing failover
+            controllerInstance.scaleService(3, true);
+            segmentStoreInstance.scaleService(3, true);
+            Thread.sleep(ZK_DEFAULT_SESSION_TIMEOUT);
+
+            //run the failover test after scaling
+            performFailoverTest();
+
+            stopReadersAndWriters(readerGroupManager, readerGroupName);
+
+        }
+        cleanUp(scope, STREAM);
+        log.info("Test {} succeeds ", "ReadWriteAndAutoScaleWithFailover");
+    }
+
+    private void waitForScaling() {
```

Comment:
It has a sleep of 60000( 1min) and wait counter iterates two times. So it waits for 2*60s = 120 s for scaling to happen.